crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,301 @@
+                ops,
+                cool_down_list,
+            })),
+            fetch_queue_tx,
+            fetch_queue_rx,
+        }
+    }
+
+    pub fn create_fetch_queue(
+        &self,
+        config: CoreFetchConfig,
+        transport: DynTransport,
+    ) {
+        for _ in 0..config.parallel_request_count {
+            tokio::task::spawn(Inner::fetch_task(
+                self.fetch_queue_tx.clone(),
+                self.fetch_queue_rx.clone(),
+                transport.clone(),
+                self.state.clone(),
+                config.cool_down_interval_ms,
+            ));
+        }
+    }
+
+    async fn fetch_task(
+        fetch_request_tx: Sender<FetchRequest>,
+        fetch_request_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+        transport: DynTransport,
+        state: Arc<Mutex<State>>,
+        cool_down_interval: u64,
+    ) {
+        while let Some((op_id, agent_id)) =
+            fetch_request_rx.lock().await.recv().await
+        {
+            // Ensure op is still in the set of ops to fetch.
+            if !state
+                .lock()
+                .await
+                .ops
+                .contains(&(op_id.clone(), agent_id.clone()))
+            {
+                continue;
+            }
+
+            // Check if agent is on cool-down list.
+            if !Inner::is_agent_cooling_down(
+                &agent_id,
+                &mut state.lock().await.cool_down_list,
+                cool_down_interval,
+            ) {
+                // Send fetch request to agent.
+                if let Err(err) = transport
+                    .lock()
+                    .await
+                    .send_op_request(op_id.clone(), agent_id.clone())
+                    .await
+                {
+                    eprintln!("could not send fetch request for op {op_id} to agent {agent_id}: {err}");
+                    state
+                        .lock()
+                        .await
+                        .cool_down_list
+                        .insert(agent_id.clone(), Instant::now());
+                }
+            }
+
+            // Re-insert the fetch request into the queue.
+            if let Err(err) = fetch_request_tx
+                .send((op_id.clone(), agent_id.clone()))
+                .await
+            {
+                eprintln!("could not re-insert fetch request for op {op_id} to agent {agent_id} in queue: {err}");
+            }
+        }
+    }
+
+    fn is_agent_cooling_down(
+        agent_id: &AgentId,
+        cool_down_list: &mut HashMap<AgentId, Instant>,
+        cool_down_interval: u64,
+    ) -> bool {
+        let a = cool_down_list.get(agent_id);
+        match a {
+            Some(instant) => {
+                let now = Instant::now();
+                if (now - *instant).as_millis() > cool_down_interval as u128 {
+                    // Cool down interval has elapsed. Remove agent from list.
+                    cool_down_list.remove(agent_id);
```

Comment:
This is side-effect-esque, but I think it's fine because it's just purging a stale entry. Otherwise I'd need to run this check again in a separate function to purge, and I didn't want to because the function would be run on every fetch request in the queue.